exasol/bucketfs/__init__.py

```diff
@@ -0,0 +1,385 @@
+"""
+This module contains python api to programmatically access exasol bucketfs service(s).
+
+
+.. attention:
+
+    If no python api is required, one can also use CLI tools like CURL and HTTPIE to access bucketfs services.
+
+    Example's using CURL and HTTPIE
+    -------------------------------
+
+    1. Listing buckets of a bucketfs service
+
+        HTTPIE:
+          $ http GET http://127.0.0.1:6666/
+
+        CURL:
+          $ curl -i http://127.0.0.1:6666/
+
+
+    2. List all files in a bucket
+    
+        HTTPIE:
+          $  http --auth w:write --auth-type basic GET http://127.0.0.1:6666/default
+
+        CURL:
+          $ curl -i -u "w:write" http://127.0.0.1:6666/default
+
+
+    3. Upload file into a bucket
+
+        HTTPIE:
+          $  http --auth w:write --auth-type basic PUT http://127.0.0.1:6666/default/myfile.txt @some-file.txt
+
+        CURL:
+          $ curl -i -u "w:write" -X PUT --binary-data @some-file.txt  http://127.0.0.1:6666/default/myfile.txt
+
+    4. Download a file from a bucket
+
+        HTTPIE:
+          $  http --auth w:write --auth-type basic --download GET http://127.0.0.1:6666/default/myfile.txt
+
+        CURL:
+          $ curl -u "w:write" --output myfile.txt  http://127.0.0.1:6666/default/myfile.txt
+"""
+import hashlib
+from collections import defaultdict
+from pathlib import Path
+from typing import BinaryIO, ByteString, Iterable, Mapping, MutableMapping, Union
+from urllib.parse import urlparse
+
+import requests
+from requests import HTTPError
+from requests.auth import HTTPBasicAuth
+
+__all__ = [
+    "Service",
+    "Bucket",
+    "MappedBucket",
+    "as_bytes",
+    "as_string",
+    "as_file",
+    "as_hash",
+]
+
+
+class BucketFsError(Exception):
+    """Error occurred while interacting with the bucket fs service."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+def _lines(response):
+    lines = (line for line in response.text.split("\n") if not line.isspace())
+    return (line for line in lines if line != "")
+
+
+def _build_url(service_url, bucket=None, path=None) -> str:
+    info = urlparse(service_url)
+    url = f"{info.scheme}://{info.hostname}:{info.port}"
+    if bucket is not None:
+        url += f"/{bucket}"
+    if path is not None:
+        url += f"/{path}"
+    return url
+
+
+class Service:
+    """Provides a simple to use api to access a bucketfs service.
+
+    Attributes:
+        buckets: lists all available buckets.
+    """
+
+    def __init__(self, url: str, credentials: Mapping[str, Mapping[str, str]] = None):
+        """Create a new Service instance.
+
+        Args:
+            url: of the bucketfs service, e.g. `http(s)://127.0.0.1:2580`.
+            credentials: a mapping containing credentials (username and password) for buckets.
+                E.g. {"bucket1": { "username": "foo", "password": "bar" }}
+        """
+        self._url = url
```

Comment:
It will use the default `http`/`https` port which also would be expected from my point of view.
What would you explicitly check regarding the `url`?

Reply:
but the default http/https port is not the default bucketfs port